### ISSUE TITLE
High Vulnerability in d3-color < v3.1.0

### ISSUE THREAD

**Issue by ben-coble-sp** [swimlane/ngx-charts#1799]
**Describe the bug**

d3-color has a high vulnerability prior to v3.1.0. Ngx-charts should upgrade d3-color to v3.1.0 to resolve the issue. https://github.com/advisories/GHSA-36jr-mh4h-2g58.

**To Reproduce**

Run `npm audit` . d3-color will get the high vulnerability. 

**Expected behavior**

- Upgrading to v3.1.0 will remove the vulnerability when running `npm audit`.

**Screenshots**

**Demo**
Provide an online demo (stackblitz, codesandbox, or similar) where the issue can be reproduced 

**ngx-charts version**
Specify the version of ngx-charts where this bug is present

All versions below v20.1.0 or any new versions that have d3-color below v3.1.0.

**Additional context**

Github advisory https://github.com/advisories/GHSA-36jr-mh4h-2g58

**Comment by yazharasu** [comment]
Any update on this?

**Comment by batebobo** [comment]
As far as I can see there is an open PR about this by dependabot - https://github.com/swimlane/ngx-charts/pull/1800. Any reasons not to merge?

**Comment by moritz89** [comment]
Additionally packages such as
- d3-interpolate
- d3-scale
- d3-transition
- d3-brush

should also be updated as they also have dependencies on d3-color

**Comment by anthonyattard** [comment]
Getting this message as well. Hopefully this will be resolved soon as there is already #1800

**Comment by Faulkner368** [comment]
I am also getting this as well and as mentioned above the dependencies also depend on affected d3-color. The PR is approved already, hopefully will get merged soon ? https://github.com/swimlane/ngx-charts/pull/1800

**Comment by marjan-georgiev** [comment]
Fixed in `20.1.1`

**Comment by 3MR1T00** [comment]
> Maybe I'm missing something here about the solution applied:
> 
> In package.json I have this:
> 
> ```json
>     "dependencies": {
>         "@angular/cdk": "^15.0.2",
>         "@angular/cli": "^15.0.3",
>         "@angular/common": "^15.0.3",
>         "@angular/compiler": "^15.0.3",
>         "@angular/core": "^15.0.3",
>         "@angular/forms": "^15.0.3",
> ...
>         "@swimlane/ngx-charts": "^20.1.2",
> ...
>     },
>     "devDependencies": {
> ...
>         "@types/d3": "^7.4.0",
> ...
>     }
> ```
> 
> And I still get the 6 high severity vulnerabilities when running `npm i -f`.
> 
> When I run the audit fix, it reports them again, and does not fix anything. Obviously the proposed --force solution would install a very old version of ngx-charts, so it would be not desirable.
> 
> `npm audit fix`
> 
> up to date, audited 1288 packages in 6s
> 
> 166 packages are looking for funding run `npm fund` for details
> 
> # npm audit report
> d3-color <3.1.0 Severity: high d3-color vulnerable to ReDoS - [GHSA-36jr-mh4h-2g58](https://github.com/advisories/GHSA-36jr-mh4h-2g58) fix available via `npm audit fix --force` Will install @swimlane/ngx-charts@6.1.0, which is a breaking change node_modules/d3-color d3-interpolate 0.1.3 - 2.0.1 Depends on vulnerable versions of d3-color node_modules/d3-interpolate @swimlane/ngx-charts >=7.0.0 Depends on vulnerable versions of d3-brush Depends on vulnerable versions of d3-interpolate Depends on vulnerable versions of d3-scale Depends on vulnerable versions of d3-transition node_modules/@swimlane/ngx-charts d3-brush 0.1.0 - 2.1.0 Depends on vulnerable versions of d3-interpolate Depends on vulnerable versions of d3-transition node_modules/d3-brush d3-scale 0.1.5 - 3.3.0 Depends on vulnerable versions of d3-interpolate node_modules/d3-scale d3-transition 0.0.7 - 2.0.0 Depends on vulnerable versions of d3-color Depends on vulnerable versions of d3-interpolate node_modules/d3-transition
> 
> 6 high severity vulnerabilities
> 
> To address issues that do not require attention, run: npm audit fix
> 
> To address all issues (including breaking changes), run: npm audit fix --force

This is because they update `d3-color` which is directed dependency of `@swimlane/ngx-charts` 
`"@swimlane/ngx-charts": {
      "version": "20.1.2",
      "resolved": "https://registry.npmjs.org/@swimlane/ngx-charts/-/ngx-charts-20.1.2.tgz",
      "integrity": "sha512-9uetDtBg++v96R9Nwgf8mvSu+6149FMQEm/jlpbcquUozfC5l+hO9i88tG3REOxnkGS10snGOCVUcPPY03nbxQ==",
      "requires": {
        "d3-array": "^2.9.1",
        "d3-brush": "^2.1.0",
        "d3-color": "^3.1.0", <<<<<< 
        "d3-format": "^2.0.0",
        "d3-hierarchy": "^2.0.0",
        "d3-interpolate": "^2.0.1",
        "d3-scale": "^3.2.3",
        "d3-selection": "^2.0.0",
        "d3-shape": "^2.0.0",
        "d3-time-format": "^3.0.0",
        "d3-transition": "^2.0.0",
        "rfdc": "^1.3.0",
        "tslib": "^2.0.0"
      }
    }`

But `d3-interpolate` as well as `d3-transition` in their place are still depending on  vulnerable  `d3-color` version
`"d3-transition": {
      "version": "2.0.0",
      "resolved": "https://registry.npmjs.org/d3-transition/-/d3-transition-2.0.0.tgz",
      "integrity": "sha512-42ltAGgJesfQE3u9LuuBHNbGrI/AJjNL2OAUdclE70UE6Vy239GCBEYD38uBPoLeNsOhFStGpPI0BAOV+HMxog==",
      "requires": {
        "d3-color": "1 - 2",
        "d3-dispatch": "1 - 2",
        "d3-ease": "1 - 2",
        "d3-interpolate": "1 - 2",
        "d3-timer": "1 - 2"
      },
      "dependencies": {
        "d3-color": {
          "version": "2.0.0",
          "resolved": "https://registry.npmjs.org/d3-color/-/d3-color-2.0.0.tgz",
          "integrity": "sha512-SPXi0TSKPD4g9tw0NMZFnR95XVgUZiBH+uUTqQuDu1OsE2zomHU7ho0FISciaPvosimixwHFl3WHLGabv6dDgQ=="
        }
      }
    }`

`"d3-interpolate": {
      "version": "2.0.1",
      "resolved": "https://registry.npmjs.org/d3-interpolate/-/d3-interpolate-2.0.1.tgz",
      "integrity": "sha512-c5UhwwTs/yybcmTpAVqwSFl6vrQ8JZJoT5F7xNFK9pymv5C0Ymcc9/LIJHtYIggg/yS9YHw8i8O8tgb9pupjeQ==",
      "requires": {
        "d3-color": "1 - 2"
      },
      "dependencies": {
        "d3-color": {
          "version": "2.0.0",
          "resolved": "https://registry.npmjs.org/d3-color/-/d3-color-2.0.0.tgz",
          "integrity": "sha512-SPXi0TSKPD4g9tw0NMZFnR95XVgUZiBH+uUTqQuDu1OsE2zomHU7ho0FISciaPvosimixwHFl3WHLGabv6dDgQ=="
        }
      }
    },`

**Comment by 3MR1T00** [comment]
This is because they updated `d3-color` which is directed dependency of `@swimlane/ngx-charts` 

```
"@swimlane/ngx-charts": {
      "version": "20.1.2",
      "resolved": "https://registry.npmjs.org/@swimlane/ngx-charts/-/ngx-charts-20.1.2.tgz",
      "integrity": "sha512-9uetDtBg++v96R9Nwgf8mvSu+6149FMQEm/jlpbcquUozfC5l+hO9i88tG3REOxnkGS10snGOCVUcPPY03nbxQ==",
      "requires": {
        "d3-array": "^2.9.1",
        "d3-brush": "^2.1.0",
        "d3-color": "^3.1.0", <<<<<<<<<<<<<<<<< 
        "d3-format": "^2.0.0",
        "d3-hierarchy": "^2.0.0",
        "d3-interpolate": "^2.0.1",
        "d3-scale": "^3.2.3",
        "d3-selection": "^2.0.0",
        "d3-shape": "^2.0.0",
        "d3-time-format": "^3.0.0",
        "d3-transition": "^2.0.0",
        "rfdc": "^1.3.0",
        "tslib": "^2.0.0"
      }
    }

```
But `d3-interpolate` as well as `d3-transition` in their place are still depending on  vulnerable  `d3-color` version
```
"d3-transition": {
      "version": "2.0.0",
      "resolved": "https://registry.npmjs.org/d3-transition/-/d3-transition-2.0.0.tgz",
      "integrity": "sha512-42ltAGgJesfQE3u9LuuBHNbGrI/AJjNL2OAUdclE70UE6Vy239GCBEYD38uBPoLeNsOhFStGpPI0BAOV+HMxog==",
      "requires": {
        "d3-color": "1 - 2",
        "d3-dispatch": "1 - 2",
        "d3-ease": "1 - 2",
        "d3-interpolate": "1 - 2",
        "d3-timer": "1 - 2"
      },
      "dependencies": {
        "d3-color": {
          "version": "2.0.0", <<<<<<<<<<<<<<<<<<<<<<<<<
          "resolved": "https://registry.npmjs.org/d3-color/-/d3-color-2.0.0.tgz",
          "integrity": "sha512-SPXi0TSKPD4g9tw0NMZFnR95XVgUZiBH+uUTqQuDu1OsE2zomHU7ho0FISciaPvosimixwHFl3WHLGabv6dDgQ=="
        }
      }
    }

```
```
"d3-interpolate": {
      "version": "2.0.1",
      "resolved": "https://registry.npmjs.org/d3-interpolate/-/d3-interpolate-2.0.1.tgz",
      "integrity": "sha512-c5UhwwTs/yybcmTpAVqwSFl6vrQ8JZJoT5F7xNFK9pymv5C0Ymcc9/LIJHtYIggg/yS9YHw8i8O8tgb9pupjeQ==",
      "requires": {
        "d3-color": "1 - 2"
      },
      "dependencies": {
        "d3-color": {
          "version": "2.0.0", <<<<<<<<<<<<<<<<<<<<<<<<
          "resolved": "https://registry.npmjs.org/d3-color/-/d3-color-2.0.0.tgz",
          "integrity": "sha512-SPXi0TSKPD4g9tw0NMZFnR95XVgUZiBH+uUTqQuDu1OsE2zomHU7ho0FISciaPvosimixwHFl3WHLGabv6dDgQ=="
        }
      }
    }
```

**Comment by daiscog** [comment]
> Fixed in `20.1.2`

@marjan-georgiev Can this be reopened?  The issue is **not** fixed in 20.1.2.  Installing 20.1.2 will still result in d3-color version 2.0.0 also being installed as it's a transitive dependency of other `d3-*` libs that `@swimlane/ngx-charts` depend on.

As described by @3MR1T00 above, the other d3 dependencies also need to be bumped to versions that depend on the fixed version of d3-color to completely fix the issue.

**Comment by aitmouhou** [comment]
The problem is still present in the version 20.1.2
[deps.dev](https://deps.dev/npm/%40swimlane%2Fngx-charts)

**Comment by samueldmeyer** [comment]
The continuing problem has been reported as https://github.com/swimlane/ngx-charts/issues/1830

**Comment by sushmeex** [comment]
Facing similar issue in our project, we are the latest ngx charts and d3 versions

<img width="445" alt="image" src="https://user-images.githubusercontent.com/123559443/228130352-dac500c9-f499-415b-b5a6-ef29f2dcdf54.png">

**Comment by diogogasparr** [comment]
Still finding vulnerabilities issues (d3 dependency) in [20.1.2](https://www.npmjs.com/package/@swimlane/ngx-charts/v/20.1.2)

**Comment by dse-copsfs** [comment]
> Still finding vulnerabilities issues (d3 dependency) in [20.1.2](https://www.npmjs.com/package/@swimlane/ngx-charts/v/20.1.2)

```
# npm audit report

d3-color  <3.1.0
Severity: high
d3-color vulnerable to ReDoS - https://github.com/advisories/GHSA-36jr-mh4h-2g58
fix available via `npm audit fix --force`
Will install @swimlane/ngx-charts@6.1.0, which is a breaking change
node_modules/d3-interpolate/node_modules/d3-color
  d3-interpolate  0.1.3 - 2.0.1
  Depends on vulnerable versions of d3-color
  node_modules/d3-interpolate
    d3-scale  0.1.5 - 3.3.0
    Depends on vulnerable versions of d3-interpolate
    node_modules/d3-scale
      @swimlane/ngx-charts  >=7.0.0
      Depends on vulnerable versions of d3-scale
      node_modules/@swimlane/ngx-charts
```

@marjan-georgiev, could you please reopen this task, so it can be visible and fixed?

**Comment by queejie** [comment]
Unfortunately, it is still an issue with 20.3.0, despite some tweaks to it:

````
# npm audit report

d3-color  <3.1.0
Severity: high
d3-color vulnerable to ReDoS - https://github.com/advisories/GHSA-36jr-mh4h-2g58
fix available via `npm audit fix --force`
Will install @swimlane/ngx-charts@12.1.0, which is a breaking change
node_modules/@swimlane/ngx-charts/node_modules/d3-interpolate/node_modules/d3-color
node_modules/d3-brush/node_modules/d3-color
node_modules/d3-transition/node_modules/d3-color
  d3-interpolate  0.1.3 - 2.0.1
  Depends on vulnerable versions of d3-color
  node_modules/@swimlane/ngx-charts/node_modules/d3-interpolate
  node_modules/d3-brush/node_modules/d3-interpolate
  node_modules/d3-transition/node_modules/d3-interpolate
    @swimlane/ngx-charts  >=7.0.0
    Depends on vulnerable versions of d3-brush
    Depends on vulnerable versions of d3-interpolate
    Depends on vulnerable versions of d3-scale
    Depends on vulnerable versions of d3-transition
    node_modules/@swimlane/ngx-charts
    d3-scale  0.1.5 - 3.3.0
    Depends on vulnerable versions of d3-interpolate
    node_modules/@swimlane/ngx-charts/node_modules/d3-scale
    d3-transition  0.0.7 - 2.0.0
    Depends on vulnerable versions of d3-color
    Depends on vulnerable versions of d3-interpolate
    node_modules/d3-transition
      d3-brush  0.1.0 - 2.1.0
      Depends on vulnerable versions of d3-interpolate
      Depends on vulnerable versions of d3-transition
      node_modules/d3-brush

6 high severity vulnerabilities
````

**Comment by B0rrA** [comment]
This problem persists on latest version 20.3.1

 === npm audit security report ===                        
                                                                                
┌──────────────────────────────────────────────────────────────────────────────┐
│                                Manual Review                                 │
│            Some vulnerabilities require your attention to resolve            │
│                                                                              │
│         Visit https://go.npm.me/audit-guide for additional guidance          │
└──────────────────────────────────────────────────────────────────────────────┘
┌───────────────┬──────────────────────────────────────────────────────────────┐
│ High          │ d3-color vulnerable to ReDoS                                 │
├───────────────┼──────────────────────────────────────────────────────────────┤
│ Package       │ d3-color                                                     │
├───────────────┼──────────────────────────────────────────────────────────────┤
│ Patched in    │ >=3.1.0                                                      │
├───────────────┼──────────────────────────────────────────────────────────────┤
│ Dependency of │ @swimlane/ngx-charts                                         │
├───────────────┼──────────────────────────────────────────────────────────────┤
│ Path          │ @swimlane/ngx-charts > d3-interpolate > d3-color             │
├───────────────┼──────────────────────────────────────────────────────────────┤
│ More info     │ https://github.com/advisories/GHSA-36jr-mh4h-2g58            │
└───────────────┴──────────────────────────────────────────────────────────────┘
┌───────────────┬──────────────────────────────────────────────────────────────┐
│ High          │ d3-color vulnerable to ReDoS                                 │
├───────────────┼──────────────────────────────────────────────────────────────┤
│ Package       │ d3-color                                                     │
├───────────────┼──────────────────────────────────────────────────────────────┤
│ Patched in    │ >=3.1.0                                                      │
├───────────────┼──────────────────────────────────────────────────────────────┤
│ Dependency of │ @swimlane/ngx-charts                                         │
├───────────────┼──────────────────────────────────────────────────────────────┤
│ Path          │ @swimlane/ngx-charts > d3-brush > d3-interpolate > d3-color  │
├───────────────┼──────────────────────────────────────────────────────────────┤
│ More info     │ https://github.com/advisories/GHSA-36jr-mh4h-2g58            │
└───────────────┴──────────────────────────────────────────────────────────────┘
┌───────────────┬──────────────────────────────────────────────────────────────┐
│ High          │ d3-color vulnerable to ReDoS                                 │
├───────────────┼──────────────────────────────────────────────────────────────┤
│ Package       │ d3-color                                                     │
├───────────────┼──────────────────────────────────────────────────────────────┤
│ Patched in    │ >=3.1.0                                                      │
├───────────────┼──────────────────────────────────────────────────────────────┤
│ Dependency of │ @swimlane/ngx-charts                                         │
├───────────────┼──────────────────────────────────────────────────────────────┤
│ Path          │ @swimlane/ngx-charts > d3-brush > d3-transition >            │
│               │ d3-interpolate > d3-color                                    │
├───────────────┼──────────────────────────────────────────────────────────────┤
│ More info     │ https://github.com/advisories/GHSA-36jr-mh4h-2g58            │
└───────────────┴──────────────────────────────────────────────────────────────┘
┌───────────────┬──────────────────────────────────────────────────────────────┐
│ High          │ d3-color vulnerable to ReDoS                                 │
├───────────────┼──────────────────────────────────────────────────────────────┤
│ Package       │ d3-color                                                     │
├───────────────┼──────────────────────────────────────────────────────────────┤
│ Patched in    │ >=3.1.0                                                      │
├───────────────┼──────────────────────────────────────────────────────────────┤
│ Dependency of │ @swimlane/ngx-charts                                         │
├───────────────┼──────────────────────────────────────────────────────────────┤
│ Path          │ @swimlane/ngx-charts > d3-transition > d3-color              │
├───────────────┼──────────────────────────────────────────────────────────────┤
│ More info     │ https://github.com/advisories/GHSA-36jr-mh4h-2g58            │
└───────────────┴──────────────────────────────────────────────────────────────┘
┌───────────────┬──────────────────────────────────────────────────────────────┐
│ High          │ d3-color vulnerable to ReDoS                                 │
├───────────────┼──────────────────────────────────────────────────────────────┤
│ Package       │ d3-color                                                     │
├───────────────┼──────────────────────────────────────────────────────────────┤
│ Patched in    │ >=3.1.0                                                      │
├───────────────┼──────────────────────────────────────────────────────────────┤
│ Dependency of │ @swimlane/ngx-charts                                         │
├───────────────┼──────────────────────────────────────────────────────────────┤
│ Path          │ @swimlane/ngx-charts > d3-brush > d3-transition > d3-color   │
├───────────────┼──────────────────────────────────────────────────────────────┤
│ More info     │ https://github.com/advisories/GHSA-36jr-mh4h-2g58
